File: sdk/python/jamjet/cloud/budget.py
from __future__ import annotations

import threading

from .exceptions import JamJetBudgetExceeded


class BudgetManager:
    """Track cumulative LLM spend and enforce a cost ceiling."""

    def __init__(self, max_cost_usd: float | None = None) -> None:
        self._max_cost_usd = max_cost_usd
        self._spent: float = 0.0
        self._lock = threading.Lock()

    def record(self, cost_usd: float) -> None:
        """Record actual spend."""
        with self._lock:
            self._spent += cost_usd

    def check_or_raise(self, estimated_cost: float = 0.0) -> None:
        """Raise JamJetBudgetExceeded if adding *estimated_cost* would exceed the limit."""
        if self._max_cost_usd is None:
            return
        with self._lock:
            if self._spent + estimated_cost > self._max_cost_usd:
                raise JamJetBudgetExceeded(spent=self._spent, limit=self._max_cost_usd)

    @property
    def spent(self) -> float:
        with self._lock:
            return self._spent

    @property
    def remaining(self) -> float | None:
        """Remaining budget in USD, or None if no limit is set."""
        if self._max_cost_usd is None:
            return None
        with self._lock:
            return max(0.0, self._max_cost_usd - self._spent)


# ---------------------------------------------------------------------------
# Module-level singleton
# ---------------------------------------------------------------------------

_budget: BudgetManager = BudgetManager()
_module_lock = threading.Lock()


def get_budget() -> BudgetManager:
    """Return the global budget manager."""
    return _budget


def set_budget(max_cost_usd: float | None) -> BudgetManager:
    """Replace the global budget manager with a new limit."""
    global _budget
    with _module_lock:
        _budget = BudgetManager(max_cost_usd=max_cost_usd)
    return _budget

File: sdk/typescript/packages/cloud/src/budget.ts
import { JamjetBudgetExceeded } from './errors.js'

export class BudgetManager {
  private _maxCostUsd: number | null
  private _spent = 0

  constructor(maxCostUsd?: number | null) {
    this._maxCostUsd = maxCostUsd ?? null
  }

  record(costUsd: number): void {
    this._spent += costUsd
  }

  checkOrThrow(opts: { estimatedCost?: number } = {}): void {
    if (this._maxCostUsd === null) return
    const projected = this._spent + (opts.estimatedCost ?? 0)
    if (projected > this._maxCostUsd) {
      throw new JamjetBudgetExceeded(this._spent, this._maxCostUsd)
    }
  }

  setLimit(maxCostUsd: number | null): void {
    this._maxCostUsd = maxCostUsd
  }

  get spent(): number {
    return this._spent
  }

  get remaining(): number | null {
    if (this._maxCostUsd === null) return null
    return Math.max(0, this._maxCostUsd - this._spent)
  }
}
